activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -144,6 +144,23 @@ def scope_association_reflection(association)
         end
     end
 
+    # +WithChain+ objects act as placeholder for queries in which +with+ does not have any parameter.
+    # In this case, +with+ can be chained to return a new relation.
+    class WithChain
+      def initialize(scope) # :nodoc:
+        @scope = scope
+      end
+
+      # Returns a new relation in which Common Table Expressions (CTEs) are flagged as recursive.
+      #
+      # See QueryMethods#with for more details.
+      def recursive(*args)
+        @scope.with_values += args
+        @scope.with_is_recursive = true
+        @scope
+      end
+    end
```

Comment:
Not sure we need this. I think `.with_recursive` is just as fine as `.with.recursive`.

Unless I'm missing something?

Reply:
Yeah, I assume you were thinking of `where.not / where.associated / where.missing`.

But in that case, I doubt we'll ever need multiple `with_`, but I see how as a enduser it might appear inconsistent :/

Reply:
I also think `.with.recursive` is more in line with existing query methods and would prefer that one.

Having said that, as @ClearlyClaire pointed out `.with_recursive`  is simpler to implement (due to missing union support in ActiveRecord) and I prefer passing relations as positional arguments.

```
Post.with.recursive(
  post_and_replies: [
    Post.where(id: 42),
    Post.joins('JOIN post_and_replies ON posts.in_reply_to_id = post_and_replies.id'),
  ]
)

# OR

Post.with_recursive(:post_and_replies, Post.where(id: 42), Post.joins('JOIN post_and_replies ON posts.in_reply_to_id = post_and_replies.id')
```

Or maybe we should add union support before implementing this ;)

Reply:
I have pushed a commit to change to `QueryMethods#with_recursive`, which works exactly the same way as `QueryMethods#with` except it also sets an instance variable to mark the `WITH` with `RECURSIVE`.

It's a separate commit so we can easily go back to `with.recursive` if that is deemed better.

> Having said that, as @ClearlyClaire pointed out `.with_recursive` is simpler to implement (due to missing union support in ActiveRecord) and I prefer passing relations as positional arguments.

I actually prefer the array because it seems to me that it is more consistent with `QueryMethods#with`, and also allows passing a single SQL snippet that does the `UNION`, should the user prefer doing that. That being said, even with positional arguments we could make the last one optional to allow that.

Reply:
:thinking: `with_recursive` seems better here, since chaining would serve only one purpose.